pennylane_forest/numpy_wavefunction.py

```diff
@@ -86,14 +82,44 @@ def var(self, observable, wires, par):
         else:
             A = observable_map[observable]
 
-        var = self.ev(A @ A, wires) - self.ev(A, wires) ** 2
+        if self.shots == 0:
+            # exact variance value
+            var = self.ev(A @ A, wires) - self.ev(A, wires) ** 2
+        else:
+            # estimate the variance
+            var = np.var(self.sample(observable, wires, par, self.shots))
+
         return var
 
+    def sample(self, observable, wires, par, n=None):
+        if n is None:
+            n = self.shots
+
+        if n == 0:
+            raise ValueError("Calling sample with n = 0 is not possible.")
+
+        if n < 0 or not isinstance(n, int):
+            raise ValueError("The number of samples must be a positive integer.")
+
+        if observable == "Hermitian":
+            A = par[0]
+        else:
+            A = observable_map[observable]
+
+        a, P = spectral_decomposition(A)
+
+        p = np.zeros(a.shape)
+
+        for idx, Pi in enumerate(P):
+            p[idx] = self.ev(Pi, wires)
+
+        return np.random.choice(a, n, p=p)
+
     def ev(self, A, wires):
```

Comment:
There's code repetition between this and the same method in `wavefunction.py`. Consider moving into `ForestDevice`.

Reply:
Addressed as per https://github.com/rigetti/pennylane-forest/pull/18#discussion_r311904647